### ISSUE TITLE
conditional sub-requests / back-end registry / openapi/raml imports

### ISSUE THREAD

**Issue by roscopecoltran** [devopsfaith/krakend#47]
Hey,

Just creating a new issue for the additional questions/suggestions I had... ^^

**1.** How would u see to trigger some conditional sub-requests (in serie or in parallel) ? the idea would be to add some flow based sub-requests per backend or endpoint.

  - **Refs**. 
       - Mixed Request Queues: https://github.com/xh3b4sd/taskq/blob/master/taskq_mixed_test.go
       - Flow Based su-requests: https://github.com/mackristof/react-webapp/blob/master/main/react-webapp.go

  - **Use case**: would be to fetch github user info while requesting a github search and some other request like twitter API for user info if a match is possible with the user info from github.

**2.** How would you see to preload some backends/providers profiles and hook them in the config files with backend_id ?

  - **Use case**: Avoid redundancy of backend configuration or to auto suggest some backend configuration based on their API definitions (RAML based or OpenAPI based)
  - **Refs**
     -  https://api.apis.guru/v2/list.json
     - https://github.com/qor/admin

Cheers,
Richard

**Comment by kpacha** [comment]
Hi, Richard

### 1. Conditional sub-requests

I'm not sure that's something you should do in the api-gateway layer because it will add complexity and latency to a layer intended to be as reactive as possible. See the [KrakenD rules](https://github.com/devopsfaith/krakend/blob/master/docs/OVERVIEW.md#the-krakend-rules) for more.

But, it's up to you to build a proxy middleware able to parse some custom config and delegate to it the responsibility of blocking the response, triggering the related pipe(s) and merge the responses of every sub-request to the original response.

Following your example, you could define 2 endpoints (named `github_search` and `twitter_search`) and the custom proxy middleware described above. You could configure the `github_search` endpoint to use the custom proxy middleware that blocks the response, triggers N http requests to the `twitter_search` and mixes the results into the original ones.

Please, notice this will add a huge overload to the engine (in terms of memory, goroutines and filedescriptors consumed per request) and I see some hard decisions to make during the implementation like: how to deal with partial errors, what's the timeout to apply, and so on

### 2. backend/providers profiles

I'm not sure if I got your question right...

If you're talking about using something like the engines of the 'searx' project, I think it would be great to get some 'out-of-the-box' plugins for the most common backends. But I'm not sure if there are such things, since almost everyone is using KrakenD to consume its own api catalogue.

If your question was more about integrating RAML or swagger or any other API description language/module/whatever, that's something we have in the roadmap and I'd love to see what do you have in mind

cheers!

PS: it would be nice to have every question on a dedicated issue, so it would be easier for the community to search for a specific one.
